onnxruntime/core/providers/qnn/builder/opbuilder/matmul_op_builder.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/providers/common.h"
+#include "core/providers/qnn/builder/op_builder_factory.h"
+#include "core/providers/qnn/builder/opbuilder/base_op_builder.h"
+#include "core/providers/qnn/builder/qnn_model_wrapper.h"
+#include "core/providers/qnn/builder/qnn_utils.h"
+#include "core/providers/shared/utils/utils.h"
+
+namespace onnxruntime {
+namespace qnn {
+
+// If MatMul's 2nd input is 2D initializer, we can transpose the 2nd input and use FullyConnected Op instead,
+// which is faster and using MatMul Op.
+class MatMulOpBuilder : public BaseOpBuilder {
+ public:
+  MatMulOpBuilder() : BaseOpBuilder("MatMulOpBuilder") {}
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(MatMulOpBuilder);
+
+ protected:
+  Status ProcessInputs(QnnModelWrapper& qnn_model_wrapper, const NodeUnit& node_unit, const logging::Logger& logger,
+                       std::vector<std::string>& input_names, bool do_op_validation) const override ORT_MUST_USE_RESULT;
+
+  Status ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper, const NodeUnit& node_unit,
+                                     std::vector<std::string>&& input_names, const logging::Logger& logger,
+                                     bool do_op_validation) const override ORT_MUST_USE_RESULT;
+};
+
+Status MatMulOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper, const NodeUnit& node_unit,
+                                      const logging::Logger& logger, std::vector<std::string>& input_names,
+                                      bool do_op_validation) const {
+  ORT_UNUSED_PARAMETER(do_op_validation);
+  const auto& inputs = node_unit.Inputs();
+
+  // Process input 0.
+  const std::string& input_0_name = inputs[0].node_arg.Name();
+  input_names.emplace_back(input_0_name);
+  if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_0_name)) {
+    LOGS(logger, VERBOSE) << "Tensor already added, skip it: " << input_0_name;
+  } else {
+    QnnTensorWrapper input_0_tensor;
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.MakeTensorWrapper(inputs[0], input_0_tensor));
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.AddTensorWrapper(std::move(input_0_tensor)), "Failed to add tensor.");
+  }
+
+  // Process input 1.
+  const std::string& input_1_name = inputs[1].node_arg.Name();
+  input_names.emplace_back(input_1_name);
+  bool is_initializer_input_1 = qnn_model_wrapper.IsInitializerInput(input_1_name);
+  std::vector<uint32_t> input_1_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[1].node_arg, input_1_shape), "Cannot get shape");
+  bool is_quantized_weight = inputs[1].quant_param.has_value();
+  // Use FullyConnected if 2nd input is 2D initializer. Need to transpose this 2D initializer.
+  // To make it simple, not doing this for quantized weight for now.
```

Comment:
Can we also handle 1D? Numpy rule is to append '1' to the dims, do the matmul as 2D, and remove the appended 1 after.

Reply:
This is good point. Actually the original code just doesn't handle 1D tensor on either side at all. I added some UTs and they just failed to run. I just refactored the code to support all possible cases in the new commit.